### PR TITLE
SQL Client Access: Shared database handling between client and server

### PR DESCRIPTION
**Pull request by ZenBre4ker**
To be able to access the sql database freely we are making a shared interface, which handles instatiating a database, getting and converting values as well as changing them.

TODO:
- [x] Implement serverside registering database and create them
- [x] Implement functionality to synchronize data between client and server
- [x] Implement requesting values
- [x] Handle different players requesting values
- [x] Implement conversion of values (LOCALLY for now!!!)
- [x] Implement changing values
- [x] Implement OnChange Callbacks
- [x] Document Functions
- [x] Add GetValue-Function on Server
- [x] Add Drop SQL table as Reset function
- [x] Add Local Per Item defaults
- [x] Implement it as a test case in one instance (maybe playermodels)
- [x] Test it!!!

Discussion:
1. In a next step I need to make sure, that values are properly converted. That could be my chance to implement this directly in the ORM-module or handle this locally for now and change it later. Depending what @Histalek thinks if he wants to tackle it himself or if I should just implement it in the ORM, as I guess @saibotk wont let me do another local conversion again 😅 
2. Also I wonder, SQL is quite sensitive with strings as commands in combination with user given keys, if I should prohibit read-access to everyone and only allow superadmins to access the library, even if I will try to make it as safe as possible without direct access to it, but do we still want to? I am unexperienced in this matter so I can only grasp what danger this could be.

### CODE REVIEW

**Comment by saibotk** [comment]
I think we should just schedule a little meeting, since i think we are developing mutliple syncing systems in TTT2 now (ttt2net) which all try to do roughly the same or could actually be combined.

**Comment by ZenBre4ker** [comment]
As this is quite huge, take your time to discuss this and give me feedback in your review. <3 Some things I want to point out inbefore:

- The F1 Menu playermodel code got bigger, but that is only because it is in the "front-end" programmers side. One goal of this module is to later move this code directly into the element creation, so that you have similar to the serverConvars the possibility to just say, database = name, item = itemName, key = keyName and the rest is handled for you
- I only implemented it for the playermodels, which is a not so important module and therefore can be tested with a bigger audience, without disturbing live-servers too much, should be more easily to hotfix if we dont find all bugs
-I am not sure if I like the official callback structure I mimicked of convars with (cvarName, oldValue, newValue) but i wanted to stick to a known structure
- For now I kept the conversion inside this module, wont be too hard to move it into orm or something like that later on, but as we have to send everything over the network anyways, having a converter on hand should be nice.
- Another goal is to centralize all our sql-activities over this module to be able to easily sync between client and server and set values over the F1-Menu, for example the item-values like AutoSpawnable or any value you want to change, that is normally saved in the weapons table on loading

In general I tried to make the module itself as modular as useful and if we want to move the syncing or conversion to another module that should be easily doable later, first I want to get the kinks out though and see if this is working as intended.

So future TODOs are:
- [ ] Use Module for all SQL-Access (e.g. items)
- [ ] Implement Database-Access into all elements like slider, checkboxes and dropdowns similar to serverConVars
- [ ] Maybe fuse ttt2net, cvars and database sync into one module
- [ ] Make better net.Write stuff similar to the streamTable method, that doesnt care about size

**Comment by ZenBre4ker** [comment]
Bump. Waiting for reviews.

**Comment by ZenBre4ker** [comment]
I just wanted to nicely remember, if we merge this, I'd start to implement proper database handling for sliders, checkboxes and so on.
Could also need this for my own addon, so I can easily use sql database without having to write a lot of code 😄 

Also this feature doesnt go deep into any structures, so errors would only affect the playermodels and should be easily debugged as I couldnt find anything anymore in my playtests.

**Comment by saibotk** [comment]
I remember, that we wanted to talk about this first, since there are some points in the general idea of this PR that i disagree with being useful. From my current overview, i do not really see a good point in implementing another synchronization method, especially when handling rather complex processes as managing databases. There are many pain points like migrations etc that a developer should decide on. 
Other than that, we already can easily sync data to players and database handling is also pretty easy. 
That is why i think we need to talk about this rather than going back and forth :)

**Comment by ZenBre4ker** [comment]
Well I thought I made my points clear, that this is not about syncing methods (ttt2net) or database handling (orm/migrations) in general, which we already have, its a simple tool to easily have acces from the client to the server database for the F1 Menu combining all features into accesible functions.

And as there is currently not much happening I would love to keep going in implementing more useful stuff for F1 Menu making access for database features easy. As those functions are modular and the internal structure can be changed, if somebody really wants to improve it, they can implement other syncing methods or add migrations later on. Right now, I only need it for client access to server database. And thought a generalized approach can be more useful later on ^^

**Comment by ZenBre4ker** [comment]
Ah yeah and I also need some form of callback functions similar to convars, so that database changes can be instantly shown on the client. And as there is no way beside hooking to sql function I though combining them in one centralized place helps.
This is mainly meant for internal stuff in ttt2.

**Comment by ZenBre4ker** [comment]
After discussion:
- [ ] Check Security Concerns (SQL Read-, Write-Access)
- [ ] Make it possible to limit acces to database (e.g. admin-only)

**Comment by ZenBre4ker** [comment]
For the accessLevel restriction you can give read access to any, admins or only server in case you want to use the library on the server too, but dont need client access. Whereas write access is hardcoded to be limited to at minimum admin or higher, such that careless programmmers cant give full access to anything.

What security concerns, I used the ORM-Model and use the functions:
orm.Make, :Find, :All, :New, :Save, :Delete and the access-operator [ ]
So as long as access per itemname is protected via orm, my wrapper should be secure as well.
You cant access any other sql table beside the ones registered by the server, which should limit abuse.

File: lua/ttt2/libraries/database.lua
Comment (TimGoll) [inline lua/ttt2/libraries/database.lua:337]: should this internal function be visble to the outside? or is it enough to have a local function do this?

File: lua/ttt2/libraries/database.lua
Comment (TimGoll) [inline lua/ttt2/libraries/database.lua:417]: add an empty line here

File: lua/ttt2/libraries/database.lua
Comment (TimGoll) [inline lua/ttt2/libraries/database.lua:425]: here an empty line as well

File: lua/ttt2/libraries/database.lua
Comment (TimGoll) [inline lua/ttt2/libraries/database.lua:550]: why this empty line here?

File: lua/ttt2/libraries/database.lua
Comment (TimGoll) [inline lua/ttt2/libraries/database.lua:1259]: remove emty line

**Review by TimGoll — APPROVED** [review]
I see the points made by @saibotk but I still approve this change because it mainly simplifies the access needed for the F1 menu and brings the settings integration forward. From my point of view this is a nice addition and I'm open for a discussion with you and @saibotk

**Review by Histalek — CHANGES_REQUESTED** [review]
Because of the comments made by Zen about the ORM usage + security concerns i really want to take a closer look at this soon.

That's why i want to prevent a merge for now

**Review by Histalek — CHANGES_REQUESTED** [review]
OK looks mostly good

A few notes:
- i'm not confident on commenting on the net message part
- security wise this doesn't look bad
	- write access restricted to at least admin is good
	- read access possible for anyone should be fine as there shouldn't be sensitive information stored in the database. And it would need someone with lua access to expose such tables anyway
	
Other than that, i have trouble understanding what you mean by this:
> So as long as access per itemname is protected via orm, my wrapper should be secure as well.

What exactly should be "protected" here?
- Access to a database table only from one 'source' ? That is not possible
- Protected from concurrent access/writes (from the point of view of the orm)? I'm not sure if i put a safeguard in place for that in the orm. Would need to look into it
- protected from 'unauthorized' access to the database table? Not really possible, as anyone with lua access (on the server or client respectively) can write to the database anyway

Would be nice to get a clarification on this :)

File: lua/ttt2/libraries/database.lua
Comment (TimGoll) [reply lua/ttt2/libraries/database.lua:113]: "... is not available or synced" is fine though, you don't have to negate it both

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:29]: TTT2_ is a good idea.
Wrongly interpreted in the way that 0 means no access? Thats possible.
As saibotk suggested that there should be more groups inbetween, I would say, that any access could be then 10 and server restricted -10, so I can easily work with greater/smaller >=< etc.

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:104]: Its actually not important, it's just that the accessName doesnt have to be the real one and you could choose a way smaller one, more easier to remember or write. Maybe the exclamation mark there is too much.

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:337]: I guess not. I will re-check why I did that,

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:636]: Well yeah I wanted to, because the code itself is not used anywhere. With the synced default values, this function is most likely never reached. So I only left a warning for edge cases that I forgot to think about and my code just keeps working normally.

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:1089]: Well I wanted it in there, because it might be, that that table was deleted and in the f1 menu it will be only registered in case its needed. So I thought instead of making another create, I just create it in case it's not there to minimize casualties.
It also makes sure that there really is a database it can call.

**Review by ZenBre4ker — COMMENTED** [review]
Concerning the security question in line https://github.com/TTT-2/TTT2/pull/894/files#diff-af66804b71e6a22f0b71a9c3ed356117a7c406369ffc10c180cf9081b334187dR1189
is about that the itemName a string could be anything. So if accessing the orm via the find function for example could lead to such that the table is being written to would be a problem.
But I checked your orm implementation and with sqlstr it seems to escape all dangerous characters anyway, so that the table is really only searched for that itemname and not maliciously changed by a well defined string on clientside.

File: lua/ttt2/libraries/database.lua
Comment (Histalek) [reply lua/ttt2/libraries/database.lua:113]: you don't have to yes, but that sentence is structurally ambiguous 
is it:
`(is not available) or (synced)`
or
`is not (available or synced)`

That's why i want to be explicit on this. But that's not a hard blocker tbh

File: lua/ttt2/libraries/database.lua
Comment (Histalek) [reply lua/ttt2/libraries/database.lua:1089]: Valid point. could we place the creation of the database behind a flag instead then?

Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:1090]: Hm not sure if that makes sense, I started to draft it and came to a point where it seems pointless. I would still need to check if the table really exists and that is the same sql.CreateSqlTable does before it creates one. Also the user would still need to give me the databaseName and the savingKeys, so my library can handle it. So I could mention that in the documentation.

Im not sure why someone would try to register a database that doesnt exist. And if it exists its not created anyways. I dont really see the point in that flag, which freedom it would deliver. Only thing I see is, that I could throw an error if the user says it exists but it doesnt.

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:337]: I guess I thought I would use it outside, but I am not. Probably still an early design idea, I threw away, I localized it 😄

**Review by Histalek — APPROVED** [review]

**Comment by TimGoll** [comment]
@saibotk just a ping for visibility

File: lua/ttt2/libraries/database.lua
Comment (saibotk) [inline lua/ttt2/libraries/database.lua:240]: This seems to be wrong, since it does not choose one, but abuses the same string.

File: lua/ttt2/libraries/database.lua
Comment (saibotk) [reply lua/ttt2/libraries/database.lua:637]: id rather remove them

**Review by saibotk — CHANGES_REQUESTED** [review]

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:240]: Just done like for convars. It doesnt need to be globally unique as well.
https://wiki.facepunch.com/gmod/cvars.AddChangeCallback

File: lua/ttt2/libraries/database.lua
Comment (ZenBre4ker) [reply lua/ttt2/libraries/database.lua:637]: Can do that ^^

**Comment by ZenBre4ker** [comment]
Okay, this should be it @saibotk. Now registeredPlayers cant be explloited, you only get the information of database you are registered to.

**Review by saibotk — APPROVED** [review]

**Comment by ZenBre4ker** [comment]
Okay, it's tested and works. Thanks for your input ❤️ gonna merge it now.

**Pull request merged**
